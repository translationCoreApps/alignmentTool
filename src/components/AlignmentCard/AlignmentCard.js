import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * Generates the styles for the component
 * @param props
 * @return {*}
 */
const makeStyles = (props) => {
  const { topWords, hoverTop, hoverBottom, bottomWords, acceptsTopWords, acceptsBottomWords } = props;
  const emptyTop = !topWords || topWords.length === 0;
  const emptyBottom = !bottomWords || bottomWords.length === 0;
  const emptyAlignment = emptyTop && emptyBottom;
  const largeAlignment = (!emptyTop && topWords.length > 1) || (!emptyBottom && bottomWords.length > 1);

  const defaultAlignmentWidth = '115px';
  const blueBorder = '3px dashed #44C6FF';
  const clearBorder = '3px dashed transparent';
  const whiteBorder = '3px dashed #ffffff';
  const transitionSpeed = '0.25s';

  const rowStyle = {
    display: 'flex',
    transition: transitionSpeed,
    position: 'relative'
  };
  const styles = {
    root: {
      transition: transitionSpeed,
      padding: '7px',
      backgroundColor: '#DCDCDC',
      margin: '0px 10px 10px 0px',
      minWidth: emptyAlignment ? `calc(${defaultAlignmentWidth}/2)` : defaultAlignmentWidth,
      flexGrow: largeAlignment ? 1 : 0
    },
    content: {
      display: 'flex',
      flexDirection: 'column',
      height: '100%',
      backgroundColor: '#DCDCDC'
    },
    top: {
      transition: transitionSpeed,
      flexGrow: 1,
      width: '100%',
      minHeight: '45px',
      border: emptyTop || acceptsTopWords ? whiteBorder : clearBorder,
      boxSizing: 'border-box',
      marginBottom: '7px'
    },
    bottom: {
      transition: transitionSpeed,
      flexGrow: 1,
      width: '100%',
      minHeight: '45px',
      border: emptyBottom || acceptsBottomWords ? whiteBorder : clearBorder,
      boxSizing: 'border-box'
    },
    topRow: {
      ...rowStyle,
      top: acceptsTopWords ? '7px' : 0,
      left: acceptsTopWords ? '7px' : 0,
      opacity: hoverTop ? '0.8' : 1
    },
    bottomRow: {
      ...rowStyle,
      top: acceptsBottomWords ? '7px' : 0,
      left: acceptsBottomWords ? '7px' : 0,
      opacity: hoverBottom ? '0.8' : 1
    }
  };

  if (hoverTop && acceptsTopWords) {
    styles.top.border = blueBorder;
  }
  if (hoverBottom && acceptsBottomWords) {
    styles.bottom.border = blueBorder;
  }
  return styles;
};

/**
 * Renders the alignment of primary and secondary words/phrases
 *
 * @property {array} topWords
 * @property {array} bottomWords
 * @property {bool} hoverBottom - a bottom word is hover over this component
 * @property {bool} hoverTop - a top word is hovering over this component
 * @property {bool} acceptsTopWords - this component accepts dropped top words
 * @property {bool} acceptsBottomWords - this component accepts dropped bottom words
 */
class AlignmentCard extends Component {
  render() {
    const { bottomWords, topWords } = this.props;
    const styles = makeStyles(this.props);

    return (
      <div style={styles.root}>
        <div style={styles.content}>
          <div style={styles.top}>
            <div style={styles.topRow}>
              {topWords}
            </div>
          </div>
          <div style={styles.bottom}>
            <div style={styles.bottomRow}>
              {bottomWords}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

AlignmentCard.propTypes = {
  topWords: PropTypes.array,
  bottomWords: PropTypes.array,
  hoverBottom: PropTypes.bool,
  hoverTop: PropTypes.bool,
  acceptsTopWords: PropTypes.bool,
  acceptsBottomWords: PropTypes.bool,
};

export default AlignmentCard;
