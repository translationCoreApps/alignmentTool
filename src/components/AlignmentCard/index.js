import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {DropTarget} from 'react-dnd';
import * as types from '../WordCard/Types';
import SecondaryToken from '../SecondaryToken';
import PrimaryToken from '../PrimaryToken';
import AlignmentCard from './AlignmentCard';
import Token from 'word-map/structures/Token';

const styles = {
  root: {
    open: {
      width: 'auto',
      display: 'inherit',
      transition: '0.5s'
    },
    closed: {
      width: '0',
      display: 'none',
      transition: '0.5s'
    }
  }
};

/**
 * Determines if a word can be dropped
 * @param dropTargetProps
 * @param dragSourceProps
 * @return {boolean}
 */
export const canDropPrimaryToken = (dropTargetProps, dragSourceProps) => {
  const emptyTarget = dropTargetProps.sourceNgram.length === 0;
  const singleTarget = dropTargetProps.sourceNgram.length === 1;
  const mergedTarget = dropTargetProps.sourceNgram.length > 1;
  const singleSource  = dragSourceProps.alignmentLength === 1;
  const mergedSource = dragSourceProps.alignmentLength > 1;
  const alignmentDelta = dropTargetProps.alignmentIndex - dragSourceProps.alignmentIndex;
  // const leftPlaceholder = dropTargetProps.placeholderPosition === 'left';  //alignmentDelta < 0;
  // const rightPlaceholder = dropTargetProps.placeholderPosition === 'right'; //alignmentDelta > 0;
  const moved = alignmentDelta !== 0;
  // const leftWord = mergedSource && dragSourceProps.wordIndex === 0;
  // const rightWord = mergedSource && dragSourceProps.wordIndex === dragSourceProps.alignmentLength - 1;

  // limit all drags to adjacent alignments
  if(Math.abs(alignmentDelta) > 1) return false;

  // single to single
  // TRICKY: make sure we've moved
  if(singleSource && singleTarget && moved) return true;

  // single to merged
  if(singleSource && mergedTarget) {
    return true;
  }

  // merged to empty
  if(mergedSource && emptyTarget) {
    if(!moved) return true;
    // TODO: need a workaround for this bug before supporting left vs right un-merging https://github.com/react-dnd/react-dnd/issues/735
    // see components/AlignmentGrid.js
    // we could potentially use the touch backend https://github.com/yahoo/react-dnd-touch-backend
    // however that would require us to render a custom drag preview and the drag performance may
    // not be as good.
    // if(!moved && leftPlaceholder && leftWord) return true;
    // if(!moved && rightPlaceholder && rightWord) return true;
  }

  return false;
};


/**
 * Renders the alignment of primary and secondary words/phrases
 */
class DroppableAlignmentCard extends Component {
  render() {
    const {
      translate,
      lexicons,
      alignmentIndex,
      canDrop,
      dragItemType,
      isOver,
      actions,
      targetNgram,
      connectDropTarget,
      sourceNgram
    } = this.props;

    const acceptsTop = canDrop && dragItemType === types.PRIMARY_WORD;
    const acceptsBottom = canDrop && dragItemType === types.SECONDARY_WORD;

    const hoverTop = isOver && acceptsTop;
    const hoverBottom = isOver && acceptsBottom;

    const emptyAlignment = sourceNgram.length === 0 && targetNgram.length === 0;

    const alignmentLength = sourceNgram.length;
    const topWordCards = sourceNgram.map((token, index) => (
      <PrimaryToken
        key={index}
        translate={translate}
        wordIndex={index}
        alignmentLength={alignmentLength}
        token={token}
        alignmentIndex={alignmentIndex}
        lexicons={lexicons}
        actions={actions}
      />
    ));
    const bottomWordCards = targetNgram.map((token, index) => (
      <SecondaryToken
        key={index}
        token={token}
        alignmentIndex={alignmentIndex}
      />
    ));

    if (emptyAlignment && !canDrop) {
      return <div style={styles.root.closed}/>;
    } else {
      return connectDropTarget(
        <div>
          <AlignmentCard targetTokenCards={bottomWordCards}
                         hoverBottom={hoverBottom}
                         hoverTop={hoverTop}
                         acceptsTargetTokens={acceptsBottom}
                         acceptsSourceTokens={acceptsTop}
                         sourceTokenCards={topWordCards}/>
        </div>
      );
    }
  }
}

DroppableAlignmentCard.propTypes = {
  translate: PropTypes.func.isRequired,
  placeholderPosition: PropTypes.string,
  dragItemType: PropTypes.string,
  isOver: PropTypes.bool.isRequired,
  canDrop: PropTypes.bool.isRequired,
  connectDropTarget: PropTypes.func.isRequired,
  sourceNgram: PropTypes.arrayOf(PropTypes.instanceOf(Token)).isRequired,
  targetNgram: PropTypes.arrayOf(PropTypes.instanceOf(Token)).isRequired,
  alignmentIndex: PropTypes.number.isRequired,
  onDrop: PropTypes.func.isRequired,
  lexicons: PropTypes.object.isRequired,
  actions: PropTypes.shape({
    showPopover: PropTypes.func.isRequired,
    loadLexiconEntry: PropTypes.func.isRequired
  })
};

const dragHandler = {
  canDrop(props, monitor) {
    const item = monitor.getItem();
    const alignmentEmpty = (props.sourceNgram.length === 0 &&
      props.targetNgram.length === 0);
    let canDrop = false;
    if (item.type === types.SECONDARY_WORD) {
      const alignmentIndexDelta = props.alignmentIndex - item.alignmentIndex;
      canDrop = alignmentIndexDelta !== 0 && !alignmentEmpty;
      return canDrop;
    }
    if (item.type === types.PRIMARY_WORD) {
      return canDropPrimaryToken(props, item);
    }
  },
  drop(props, monitor) {
    props.onDrop(monitor.getItem());
  }
};

const collect = (connect, monitor) => {
  const item = monitor.getItem();
  return {
    connectDropTarget: connect.dropTarget(),
    dragItemType: item ? item.type : null,
    isOver: monitor.isOver(),
    canDrop: monitor.canDrop()
  };
};

export default DropTarget(
  [types.SECONDARY_WORD, types.PRIMARY_WORD],
  dragHandler,
  collect
)(DroppableAlignmentCard);
