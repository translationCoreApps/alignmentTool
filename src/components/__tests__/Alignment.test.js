import React from 'react';
import renderer from 'react-test-renderer';
import AlignmentCard from '../AlignmentCard';
import TestBackend from 'react-dnd-test-backend';
import {DragDropContext} from 'react-dnd';

test('empty alignment', () => {
  const ConnectedAlignment = wrapInTestContext(AlignmentCard);
  const wrapper = renderer.create(
    <ConnectedAlignment topWords={[]}
                        bottomWords={[]}
                        onDrop={jest.fn()}
                        lexicons={{}}
                        alignmentIndex={0}/>
  );
  expect(wrapper).toMatchSnapshot();
});

test('filled alignment', () => {
  const ConnectedAlignment = wrapInTestContext(AlignmentCard);
  const wrapper = renderer.create(
    <ConnectedAlignment topWords={[]}
                        bottomWords={[]}
                        onDrop={jest.fn()}
                        lexicons={{}}
                        alignmentIndex={0}/>
  );
  expect(wrapper).toMatchSnapshot();
});


// TODO: test with something in it.


/**
 * Wraps a component into a DragDropContext that uses the TestBackend.
 */
function wrapInTestContext(DecoratedComponent) {
  return DragDropContext(TestBackend)(
    class TestContextContainer extends React.Component {
      render() {
        return <DecoratedComponent {...this.props} />;
      }
    }
  );
}
