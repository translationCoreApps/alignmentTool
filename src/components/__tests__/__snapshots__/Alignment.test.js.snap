// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty alignment 1`] = `null`;

exports[`filled alignment 1`] = `null`;

exports[`snapshot 1`] = `<div />`;
